### PR TITLE
feature(postgres): Define session role for connection

### PR DESCRIPTION
**Pull request by rldhont**
# Description

Be able to define a session role for postgres connection.

Documentation: https://www.postgresql.org/docs/current/sql-set-role.html

The `session_role` setting will be used to set the current user identifier
of the current SQL session to be `session_role`. Permissions checking for
SQL commands is carried out as though the named role were the one that had
logged in originally.

This is useful to automatically give the ownership of a new object (Table, View, Function) to
the `session_role` group and thus share ownership and associated rights with all members 
of the `session_role` group.

The specified `session_role` must be a role that the current session user
is a member of. (If the session user is a superuser, any role can be selected.)

Funded by 3liz https://3liz.com/

### CODE REVIEW

**Comment by nyalldawson** [comment]
LGTM! Is it possible to test this?

File: tests/src/providers/testqgspostgresconn.cpp [hunk 1]
```diff
@@ -203,6 +204,43 @@ class TestQgsPostgresConn: public QObject
...
+
+      uri.setParam( QStringLiteral( "session_role" ),  QStringLiteral( "qgis_test_group" ) );
+      conn = QgsPostgresConn::connectDb( uri.connectionInfo( true ), true );
+      QVERIFY( conn != 0 );
+      QgsPostgresResult result = conn->PQexec( sql );
+      QCOMPARE( result->PQgetvalue( 0, 0 ), "qgis_test_user" );
+      QCOMPARE( result->PQgetvalue( 0, 1 ), "qgis_test_group" );
+      conn->unref();
+
+      uri.setParam( QStringLiteral( "session_role" ),  QStringLiteral( "qgis_test_unkown_group" ) );
```
Comment (lbartoletti) [inline tests/src/providers/testqgspostgresconn.cpp:231]:
```suggestion
      uri.setParam( QStringLiteral( "session_role" ),  QStringLiteral( "qgis_test_unknown_group" ) );
```

**Review by lbartoletti — APPROVED** [review]
lgtm

File: tests/src/providers/testqgspostgresconn.cpp [hunk 2]
```diff
@@ -203,6 +204,43 @@ class TestQgsPostgresConn: public QObject
...
     }
+
+    void sessionRole()
+    {
+      QgsPostgresConn *conn = getConnection();
+      QVERIFY( conn != 0 );
+
+      const QString sql = QStringLiteral( "SELECT SESSION_USER, CURRENT_USER;" );
+
+      QgsPostgresResult result( conn->PQexec( sql ) );
```
Comment (nyalldawson) [inline tests/src/providers/testqgspostgresconn.cpp:215]:
```suggestion
      QgsPostgresResult result = conn->PQexec( sql );
```

File: tests/src/providers/testqgspostgresconn.cpp [hunk 3]
```diff
@@ -203,6 +204,43 @@ class TestQgsPostgresConn: public QObject
...
+
+      uri.setParam( QStringLiteral( "session_role" ),  QStringLiteral( "qgis_test_unknown_group" ) );
+      conn = QgsPostgresConn::connectDb( uri.connectionInfo( true ), true );
+      QVERIFY( conn == 0 );
+      conn->unref();
+
+      uri.removeParam( QStringLiteral( "session_role" ) );
+      conn = QgsPostgresConn::connectDb( uri.connectionInfo( true ), true );
+      QVERIFY( conn != 0 );
+      QgsPostgresResult result( conn->PQexec( sql ) );
```
Comment (nyalldawson) [inline tests/src/providers/testqgspostgresconn.cpp:239]:
```suggestion
      QgsPostgresResult result = conn->PQexec( sql );
```

File: tests/src/providers/testqgspostgresconn.cpp [hunk 4]
```diff
@@ -203,6 +204,43 @@ class TestQgsPostgresConn: public QObject
...
+      QCOMPARE( result.PQgetvalue( 0, 1 ), "qgis_test_user" );
+
+      const char *connstring = getenv( "QGIS_PGTEST_DB" );
+      if ( NULL == connstring ) connstring = "service=qgis_test";
+      QgsDataSourceUri uri( connstring );
+
+      uri.setParam( QStringLiteral( "session_role" ),  QStringLiteral( "qgis_test_group" ) );
+      conn = QgsPostgresConn::connectDb( uri.connectionInfo( true ), true );
+      QVERIFY( conn != 0 );
+      QgsPostgresResult result( conn->PQexec( sql ) );
```
Comment (nyalldawson) [inline tests/src/providers/testqgspostgresconn.cpp:226]:
```suggestion
      QgsPostgresResult result = conn->PQexec( sql );
```

File: tests/src/providers/testqgspostgresconn.cpp [hunk 5]
```diff
@@ -203,6 +204,43 @@ class TestQgsPostgresConn: public QObject
...
+      QVERIFY( conn != 0 );
+
+      const QString sql = QStringLiteral( "SELECT SESSION_USER, CURRENT_USER;" );
+
+      QgsPostgresResult result = conn->PQexec( sql );
+      QCOMPARE( result.PQgetvalue( 0, 0 ), "qgis_test_user" );
+      QCOMPARE( result.PQgetvalue( 0, 1 ), "qgis_test_user" );
+
+      const char *connstring = getenv( "QGIS_PGTEST_DB" );
+      if ( NULL == connstring ) connstring = "service=qgis_test";
```
Comment (nyalldawson) [inline tests/src/providers/testqgspostgresconn.cpp:220]: Can you fix up all the formatting and old style code here? 
Eg this should be:

```suggestion
      if ( !connstring )
          connstring = "service=qgis_test";
```

File: tests/src/providers/testqgspostgresconn.cpp [hunk 6]
```diff
@@ -203,6 +204,43 @@ class TestQgsPostgresConn: public QObject
...
+      QgsDataSourceUri uri( connstring );
+
+      uri.setParam( QStringLiteral( "session_role" ),  QStringLiteral( "qgis_test_group" ) );
+      conn = QgsPostgresConn::connectDb( uri.connectionInfo( true ), true );
+      QVERIFY( conn != 0 );
```
Comment (nyalldawson) [inline tests/src/providers/testqgspostgresconn.cpp:225]:
```suggestion
      QVERIFY( conn );
```

etc

File: src/core/qgsdatasourceuri.h [hunk 7]
```diff
@@ -404,6 +404,8 @@ class CORE_EXPORT QgsDataSourceUri
     QString mSrid;
     //! Generic params store
     QMultiMap<QString, QString> mParams;
+    //! Connection info parameters stored as generic params
+    QStringList mConnectionInfoParams = { "session_role" };
```
Comment (nyalldawson) [inline src/core/qgsdatasourceuri.h:408]: This is an ugly approach -- why is it needed here?

Comment (rldhont) [reply src/core/qgsdatasourceuri.h:408]: I need to add `session_role` parameter to `connectionInfo` string. 

I choose this approach to only update 2 methods and to simplify defined new connection information parameters: 
* `connectionInfo` to include `session_role` parameter in the conection information
* `uri` to do not duplicate `session_role` parameter in the uri string

An other approach could be to defined a private class property `mSessionRole` like `mAuthConfigId`, `mUsername` or `mPassword`. With a new private class property,  I have to update 4 methods:
* `setParam`
* `hasParam`
* `param`
* `connectionInfo`

@nyalldawson If you think defined a new private class property is better, I can update my code.

**Comment by rldhont** [comment]
The failing test is not related (SSL network access).

Hi @nyalldawson, do you want me to remove mConnectionInfoParams from QgsDataSourceUri and use a private attribute mSessionRole?

Hi @lbartoletti, I asked you to redo a code review because I had to modify the QgsDataSourceUri class.

File: src/core/qgsdatasourceuri.h (hunk 7 above)
Comment (rldhont) [reply src/core/qgsdatasourceuri.h]: I have removed the ugly code.

**Comment by nyalldawson** [comment]
I don't think session role handling belongs in QgsDataSourceUri at all -- that's a generic class, and you're putting a postgres specific property in it. That's why there's already the custom parameter API in QgsDataSourceUri. Can't this be used instead of the new member?

**Comment by rldhont** [comment]
@nyalldawson ok, I understood, I was not sure how to do it without duplicating too much code. I will remove the changes on QgsDataSourceUri and update the Postgres provider.

**Comment by nyalldawson** [comment]
Thanks @rldhont !

File: src/providers/postgres/qgspostgresconn.h [hunk 8]
```diff
@@ -211,13 +211,35 @@ class QgsPostgresConn : public QObject
...
+     * \param connInfo the QgsDataSourceUri connection info with username / password
+     * \param readOnly is the connection read only ?
      * \param shared allow using a shared connection. Should never be
      *        called from a thread other than the main one.
      *        An assertion guards against such programmatic error.
+     * \param transaction is the connection a transaction ?
+     *
+     * \returns the PostgreSQL connection
      */
     static QgsPostgresConn *connectDb( const QString &connInfo, bool readOnly, bool shared = true, bool transaction = false );
```
Comment (nyalldawson) [inline src/providers/postgres/qgspostgresconn.h:227]: Is this still used anywhere? If not, you can safely remove it -- this class isn't stable API

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
LGTM ! Thanks @rldhont

File: src/providers/postgres/qgspostgresconn.h (hunk 8 above)
Comment (rldhont) [reply src/providers/postgres/qgspostgresconn.h:227]: It is still used by QgsPostgresTransaction builder derived from QgsTransaction.

**Comment by posikifi** [comment]
Hi!

I tried to use this "Session ROLE" in QGIS 3.32.2 and I think "Session ROLE"-value is not stored to QGIS profile settings (aka QGIS3.ini).

When I add "Session ROLE"-value in PostGIS connection settings and test connection, it works as expected:

image

When I close "Connection" windows and reopen it, "Session ROLE" value is empty:

image

I expect that this "Session ROLE" -value should be stored to QGIS3.ini, but I didn't found anything: 

image

P

**Pull request merged**